### PR TITLE
Bugs, cleanup, testing

### PR DESCRIPTION
**Pull request by chris-smith**
-Moves examples into example folder
-Updates message generation/searching to use
 new version of ros_msg_utils
-Validate service client headers
-Remove portscanner dep
-Use alternate master port in tests
-Add some service tests
-Default to ROS_IP and ROS_HOSTNAME
 environment variables if provided
-Handle service servers shutting down
 between during service calls

### CODE REVIEW

**Comment by chris-smith** [comment]
Depends on [RethinkRobotics-opensource/ros_msg_utils#2](https://github.com/RethinkRobotics-opensource/ros_msg_utils/pull/2)

@b3devs @rudicus @rethink-rlinsalata

File: lib/RosNode.js [hunk 1]
```diff
@@ -295,51 +295,43 @@ class RosNode extends EventEmitter {
     return this._exit();
   }
 
-  _setupSlaveApi() {
+  _setupSlaveApi(xmlrpcPort=null) {
+    if (xmlrpcPort = null) {
```
Comment (b3devs) [inline lib/RosNode.js]: The 'if' should be === ?

File: utils/network_utils.js [hunk 2]
```diff
@@ -18,15 +18,10 @@
...
-let USE_ROS_ENV_VARS = false;
-let MIN_PORT = 49152;
-let MAX_PORT = 65535;
-
-let getRandomPort = function() {
-  return Math.round(Math.random() * (MAX_PORT - MIN_PORT) + MIN_PORT);
-}
+const ROS_IP = process.env.ROS_IP;
+const ROS_HOSTNAME = process.env.ROS_HOSTNAME;
+const HOST = ROS_IP || ROS_HOSTNAME || os.hostname();
```
Comment (rudicus) [inline utils/network_utils.js:24]: Huzzah!

**Review by rudicus — APPROVED** [review]

File: lib/ServiceClient.js [hunk 3]
```diff
@@ -172,44 +189,66 @@ class ServiceClient extends EventEmitter {
...
-    call.serviceClient.$deserializeStream = deserializer;
-    call.serviceClient.pipe(deserializer);
+        call.serviceClient.write(serviceClientHeader);
+      });
 
-    call.serviceClient.on('end', () => {
-      call.serviceClient = null;
-      // we could probably just always reset this._serviceClient to null here but...
+      // if this is a persistent service client, we're here because we haven't connected to this service before
+      // cache the service client for later use. Future calls won't need to lookup the service with the ROS master
```
Comment (b3devs) [inline lib/ServiceClient.js]: 'cache' starts a new sentence, right? Not a continuation of previous line?

File: utils/tcpros_utils.js [hunk 4]
```diff
@@ -245,6 +245,21 @@ let TcprosUtils = {
...
 
+  validateServiceClientHeader(header, service, md5sum) {
+    if (!header.hasOwnProperty('service')) {
+      return 'Connection header missing expected field [service]';
+    }
+    else if (!header.hasOwnProperty('md5sum')) {
+      return 'Connection header missing expected field [md5sum]';
+    }
+    else if (header.service !== service) {
+      return 'Got incorrect service [' + header.service + '] expected [' + service + ']';
```
Comment (b3devs) [inline utils/tcpros_utils.js:256]: In ES6, you can compose the string inline:  `Got incorrect service [${header.service}] expected [${service}]`

**Review by b3devs — APPROVED** [review]

File: lib/RosNode.js (hunk 1 above)
Comment (chris-smith) [reply lib/RosNode.js]: yep

File: lib/ServiceClient.js (hunk 3 above)
Comment (chris-smith) [reply lib/ServiceClient.js]: yep - I'll add a '.'

**Pull request merged**